text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,245 @@
+- Feature Name: cargo_stdlib_awareness
+- Start Date: 2015-05-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+
+# Summary
+
+Currently, Cargo doesn't know what parts of the standard library packages depend on.
+By giving it this knowledge, we can make cross compilation and exotic platform development easier, simplify rustbuild, and allow anyone to easily specify bounds on the standard library for their packages.
+This will allow building parts of the standard library from source, but in order to not disrupt existing workflows, the binaries that come with rustc will still be used by default.
+
+
+# Motivation
+
+First, some background.
+`rustc` can be given crates and their location with `--extern <name>=<path>`.
+When finding a library, `rustc` first sees if its location has been specified with `--extern`, then looks in any directories specified with `-L`, and finally looks in the "sysroot" [specifically `<sysroot>/lib/rustlib/<target>/lib`].
+Cargo passes in all dependencies it builds with `--extern`.
+However Cargo does not know about the standard library, so builds of it are taken from the sysroot.
+
+For cross-compiling, one can often download standard library binaries with rustup.
+This is convenient, but one cannot expect pre-built binaries for all platforms.
+In particular, embedded systems often have detailed configurations to convey as much information as possible about the hardware to the compiler.
+Furthermore, not all of the stdlib is available on every platform---there is an RFC in the works to pre-build a smaller set of crates for [Cortex-M microcontrollers](https://github.com/rust-lang/rfcs/pull/1645).
+It would be nice to know if the available subset is adequate before attempting a build.
+We can only do that if all packages have explicit standard library deps to cross-reference with platform requirements of each standard library crate.
+
+Now rustup could be augmented to build stdlib binaries in addition to downloading them, but we have extra configuration options for the standard library such as [panic strategies](https://github.com/rust-lang/rfcs/blob/master/text/1513-less-unwinding.md) along with plans to add more such as [`core` without floats](https://github.com/rust-lang/rfcs/issues/1364).
+One could also cobble together a way to to tell rustup such configuration options, but we'd like to make sure that all dependencies agree with the plan before trying to execute it.
+For panicking, packages can already set a `profile.dev.panic` option to require a specific strategy, and for float support we should add a (default opt-in) feature to core.
+Then, if packages can explicitly depend on core they can also specify whether the float feature is needed.
```

Comment:
Cargo assumes that more features can always be enabled—there is no way to require a feature to be disabled. (`default-features = false` removes the default enabled requitement, rather than adding an implicit disabled requirement.)

To be clear, I think you are critiquing more how the standard library is architected rather than Cargo should work. If we are worried about conflicting festures fragmenting the ecosystem, then we simply not use it.

For what it's eorth, I expect most would-be stdlib features will instead be handled by the scenarios system proposed in internals (on phone, no link). They'll all basically be removing functionally that the underlying hardware doesn't provided (FS, Net, atomics, etc, not just float). Another example that doesn't relate to hardware but does to compilation options is aborting panic—e.g. making `PanicError` uninhabited.